doc/api/n-api.md

```diff
@@ -0,0 +1,3024 @@
+JavaScript values. Concepts and operations generally map to ideas specified 
+in the ECMA262 Language Specification. The APIs have the following 
+properties:
+- All N-API calls return a status code of type `napi_status`. This
+  status indicates whether the API call succeeded or failed.
+- The API's return value is passed via an out parameter.
+- All JavaScript values are abstracted behind an opaque type named 
+  `napi_value`.
+- In case of an error status code, additional information can be obtained 
+  using `napi_get_last_error_info`. More information can be found in the error
+  handling section [Error Handling][].
+
+The documentation for N-API is structured as follows:
+
+* [Basic N-API Data Types][]
+* [Error Handling][]
+* [Object Lifetime Management][]
+* [Module Registration][]
+* [Working with JavaScript Values][]
+* [Working with JavaScript Values - Abstract Operations][]
+* [Working with JavaScript Properties][]
+* [Working with JavaScript Functions][]
+* [Object Wrap][]
+* [Aynchronous Operations][]
+
+The N-API is a C API that ensures ABI stability
+across Node.js versions and different compiler levels. However, we also
+understand that a C++ API can be easier to use in many
+cases. To support these cases we expect there to be one or more
+C++ wrapper modules that provide an inlineable C++ API. Binaries built
+with these wrapper modules will depend on the symbols for the N-API C based
+functions exported by Node.js. These wrappers are not part of N-API, nor
+will they be maintained as part of Node.js. One such example is:
+[node-api](https://github.com/nodejs/node-api).
+
+## Basic N-API Data Types
+
+N-API exposes the following fundamental datatypes as abstractions that are 
+consumed by the various APIs. These APIs should be treated as opaque, 
+introspectable only with other N-API calls.
+
+### *napi_status*
+Integral status code indicating the success or failure of a N-API call.
+Currently, the following status codes are supported.
+```C
+  napi_ok,
+  napi_invalid_arg,
+  napi_object_expected,
+  napi_string_expected,
+  napi_function_expected,
+  napi_number_expected,
+  napi_boolean_expected,
+  napi_generic_failure,
+  napi_pending_exception
+```
+If additional information is required upon an API returning a failed status, 
+it can be obtained by calling `napi_get_last_error_info`.
+
+### *napi_extended_error_info*
+#### Definition
+```C
+typedef struct {
+  const char* error_message;
+  void* engine_reserved;
+  uint32_t engine_error_code;
+  napi_status error_code;
+} napi_extended_error_info;
+```
+
+#### Members
+- error_message: UTF8-encoded string containing a VM-neutral description of 
+  the error
+- engine_reserved: Reserved for VM-specific error details. This is currently
+  not implemented for any VM.
+- engine_error_code: VM-specific error code. This is currently
+  not implemented for any VM.
+- error_code: the N-API status code that originated with the last error
```

Comment:
```
- `member`: Regular grammar rules.
```

Ditto for the parameters section of most APIs. I think this is the style we use elsewhere.